File: source java/virtualworld/World.java
package virtualworld;

import Interfaces.Refreshable;
import Interfaces.Requirements;
import java.awt.Color;
import java.awt.Dimension;
import java.awt.Graphics;
import java.awt.Graphics2D;
import java.awt.RenderingHints;
import java.awt.event.MouseEvent;
import java.awt.event.MouseListener;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.HashMap;
import javax.swing.BoxLayout;
import javax.swing.JComponent;
import javax.swing.JOptionPane;
import javax.swing.JPanel;
import javax.swing.SwingUtilities;
import organism.Organism;
import organism.animals.Animal;
import organism.manager.OrganismManager;
import organism.manager.Organisms;
import utilities.MyRandom;
import utilities.position.Hex;
import utilities.position.Point;
import utilities.position.Position;
import utilities.savingsystem.Loader;
import utilities.savingsystem.Saver;

public final class World 
{
    private static final int LOGS_LENGTH = 27;
    
    private static final int BOARD_X = 30;
    private static final int BOARD_Y = 30;
    private static final int TILE_SIZE = 30;
    private static final int HEX_SIZE = 18;
    private static final int BOARD_WIDTH = 700;
    private static final int BOARD_HEIGHT = 700;
    
    private int width_;
    private int height_;
    private WorldMode worldMode_;
    private final HashMap<Position, Organism> organismsPO = new HashMap<>();
    private final HashSet<Organism> RemovalSet = new HashSet<>();
    private final OrganismManager manager = new OrganismManager();
    private final LogSystem logSystem_ = new LogSystem(LOGS_LENGTH);
    
    private Refreshable refreshable_;
    
    public World(int width, int height, WorldMode worldMode)
    {
       Reset(width, height, worldMode);
    }
    
    public void SetRefreshable(Refreshable refreshable)
    {
        refreshable_ = refreshable;
    }
    
    public void Act()
    {
        ArrayList<Organism> organisms = GetElementsSortedByInitiative();
        
        for(Organism organism : organisms)
        {
            
            if(!organism.isAlive() || RemovalSet.contains(organism))
                continue;
            organism.Act();
        }
        
        EmptyRemovalSet();
        
        if(refreshable_ != null)
            refreshable_.Refresh();
    }
    
    public void Reset(int width, int height, WorldMode worldMode)
    {
        organismsPO.clear();
        width_ = width;
        height_ = height;
        worldMode_ = worldMode;
        logSystem_.Clear();
        AddToLogs("Zapis akcji:");
    }
    
    public boolean IsInBoundary(Position position)
    {
        return position.IsInBoundaries(width_, height_);
    }
    
    public Position FindNearestPosition(Position position, Requirements req)
    {
        Position result = null;
        int lowestDistance = Integer.MAX_VALUE;
        
        for(int i = 0; i < height_; i++)
            for(int j = 0; j < width_; j++)
            {
                Position tmpPosition;
                if(worldMode_ == WorldMode.SQUARE)
                    tmpPosition = new Point(j, i);
                else
                    tmpPosition = new Hex(j, i);
                
                if(tmpPosition.equals(position))
                    continue;
                
                int currentDistance = position.Distance(tmpPosition);
                if(req.IsOk(organismsPO.get(tmpPosition)) && currentDistance < lowestDistance)
                {
                    result = tmpPosition;
                    lowestDistance = currentDistance;
                }
            }
        
        return result;
    }
    
    public Position GetRandomPlaceInNeighbourhood(Position position, Requirements req)
    {
        Position NearestPosition = FindNearestPosition(position, req);
        if(NearestPosition == null || NearestPosition.Distance(position) > 1)
            return null;
        
        
        MyRandom generator = new MyRandom();
        Position result;   
        do
        {
            result = position.GetRandomNeighbour();
        }while(!IsInBoundary(result) || !req.IsOk(organismsPO.get(result)) || result.equals(position));
        
        return result;
    }
    
    private void MoveInternally(Organism organism, Position position)
    {
        Position organismOldPosition = organism.GetPosition();
        if(organismOldPosition.equals(position))
            return;
        organism.SetPosition(position);
        organismsPO.remove(organismOldPosition);
        organismsPO.remove(position);
        organismsPO.put(position, organism);
    }
    
    public void MoveAnimal(Animal animal, Position position)
    {
        if(!IsInBoundary(position))
        {
            System.out.println("Trying to move animal to place out of this world.");
            return;
        }
        
        if(organismsPO.get(position) == null || (animal.RunCollision(organismsPO.get(position)) && animal.isAlive()))
            MoveInternally(animal, position);   
    }
    
    public void SpawnOrganisms(int organisms)
    {
        MyRandom generator = new MyRandom();
        
        
        for(int i = 0; i < organisms; i++)
        {  
            Position position;
            if(worldMode_ == WorldMode.SQUARE)
                position = new Point();
            else
                position = new Hex();
            
            do
                position.SetRandom(width_, height_);
            while(organismsPO.get(position) != null);
            
            if(i == 0)
                organismsPO.put(position, manager.CreateOrganism(Organisms.HUMAN, this, position));
            else
            {
                Organisms orgEnum;
                do
                    orgEnum = Organisms.GetRandomOrganism();
                while(orgEnum == Organisms.HUMAN);
                
                Organism newOrganism = manager.CreateOrganism(orgEnum, this, position);
                organismsPO.put(position, newOrganism);
            }
        }
    }
    
    public void AddToRemovalSet(Organism organism)
    {
        RemovalSet.add(organism);
        Position position = organism.GetPosition();
        organismsPO.remove(position);
    }
    
    private void EmptyRemovalSet()
    {
        RemovalSet.clear();
    }
    
    public JPanel GetJPanel()
    {
        JPanel panel = new JPanel();
        panel.setLayout(new BoxLayout(panel, BoxLayout.X_AXIS));
        
        panel.add(GetJComponent());
        panel.add(logSystem_.GetJComponent());
        
        return panel;
    }
    
    private JComponent GetJComponent()
    {
        JComponent component = new JComponent()
        {
            @Override
            protected void paintComponent(Graphics g)
            {
                
                Graphics2D g2d = (Graphics2D) g;
                g2d.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);
                
                if(worldMode_ == WorldMode.SQUARE)
                {
                    for(int i = 0; i<height_; i++)
                        for(int j = 0; j<width_; j++)
                        {
                            Color color = Color.LIGHT_GRAY;
                            String text = "";
                            Organism organism = organismsPO.get(new Point(j, i));

                            if(organism != null)
                            {
                                color = organism.GetColor();
                                text = organism.toString().substring(0, 2);
                            }

                            g2d.setColor(color);
                            g2d.fillRect(BOARD_X + j * TILE_SIZE, BOARD_Y + i * TILE_SIZE, TILE_SIZE, TILE_SIZE);
                            g2d.setColor(Color.GRAY);
                            g2d.drawRect(BOARD_X + j * TILE_SIZE, BOARD_Y + i * TILE_SIZE, TILE_SIZE, TILE_SIZE);
                            g2d.drawString(text, BOARD_X + j * TILE_SIZE + 7, BOARD_Y + i * TILE_SIZE + g2d.getFontMetrics().getHeight() + 4);
                        }
                }
                else
                {
                    for(int i = 0; i<height_; i++)
                        for(int j = 0; j<width_; j++)
                        {
                            Color color = Color.LIGHT_GRAY;
                            String text = "";
                            Organism organism = organismsPO.get(new Hex(j, i));
                            
                            if(organism != null)
                            {
                                color = organism.GetColor();
                                text = organism.toString().substring(0, 2);
                            }
                            
                            g2d.setColor(color);
                            
                            int w = (int) ((int) HEX_SIZE * Math.sqrt(3));
                            int h = HEX_SIZE * 2;
                            int xShift = BOARD_X + j*w + (i%2 == 1 ? w/2 : 0);
                            int yShift = BOARD_Y + i*3*h/4;
                            
                            int[] xPoints = new int[6];
                            int[] yPoints = new int[6];
                            xPoints[0] = -w/2;
                            yPoints[0] = -h/4;
                            xPoints[1] = 0;
                            yPoints[1] = -h/2;
                            xPoints[2] = w/2;
                            yPoints[2] = -h/4;
                            xPoints[3] = w/2;
                            yPoints[3] = h/4;
                            xPoints[4] = 0;
                            yPoints[4] = h/2;
                            xPoints[5] = -w/2;
                            yPoints[5] = h/4;
                            for(int k = 0; k < 6; k++)
                            {
                                yPoints[k] += yShift;
                                xPoints[k] += xShift;
                            }
                            
                            g2d.fillPolygon(xPoints, yPoints, 6);
                            g2d.setColor(Color.GRAY);
                            g2d.drawPolygon(xPoints, yPoints, 6);
                            g2d.drawString(text, xShift - 8, yShift + 3);
                            
                        }
                }
            }
        };
        
        component.setPreferredSize(new Dimension(BOARD_WIDTH, BOARD_HEIGHT));
        component.addMouseListener(new MouseListener(){
            @Override
            public void mouseClicked(MouseEvent e){
                
                Position position = GetPositionByScreenCoord(e.getPoint());
                if(!IsInBoundary(position))
                    return;
                
                System.out.println(position);
                if(organismsPO.get(position) == null)
                {
                    Organisms orgEnum = (Organisms)JOptionPane.showInputDialog(
                            SwingUtilities.getWindowAncestor(component),
                            "Wybierz organizm do stworzenia:",
                            "Nowy organizm",
                            JOptionPane.PLAIN_MESSAGE,
                            null,
                            Organisms.GetValuesMinusHuman(), 
                            Organisms.GetValuesMinusHuman()[0]);
                    
                    if(orgEnum == null)
                        return;
                    
                    CreateOrganism(orgEnum, position);
                    if(refreshable_ != null)
                        refreshable_.Refresh();
                }
            }
            
            @Override
            public void mousePressed(MouseEvent e) {}
            @Override
            public void mouseReleased(MouseEvent e) {}
            @Override
            public void mouseEntered(MouseEvent e) {}
            @Override
            public void mouseExited(MouseEvent e) {}
            
        });
        return component;
    }
    
    private ArrayList<Organism> GetElementsSortedByInitiative()
    {
        ArrayList<Organism> result = new ArrayList(organismsPO.values());
        result.sort(Organism.OrganismInitiativeComparator);
        return result;
    }
    
    public void CreateOrganism(Organisms orgEnum, Position position)
    {
        Organism newOrganism = manager.CreateOrganism(orgEnum, this, position);
        organismsPO.put(position, newOrganism);
    }
    
    public void Save(Saver saver)
    {
        saver.InsertPrefix("#WIDTH");
        saver.Print(Integer.toString(width_));
        saver.InsertPrefix("#HEIGHT");
        saver.Print(Integer.toString(height_));
        saver.InsertPrefix("#WORLDMODE");
        saver.Print(worldMode_.toString());
        saver.InsertPrefix("#ORGANISMS_NUMBER");
        saver.Print(Integer.toString(organismsPO.size()));
        
        saver.InsertPrefix("#ORGANISMS");
        for(int i = 0; i<height_; i++)
            for(int j = 0; j<width_; j++)
            {
                Position position;
                if(worldMode_ == WorldMode.SQUARE)
                    position = new Point(j, i);
                else
                    position = new Hex(j, i);
                
                if(organismsPO.get(position) == null)
                    continue;
                
                position.Save(saver);
                manager.SaveOrganism(organismsPO.get(position), saver);
                    
            }
    }
    
    public void Load(Loader loader)
    {
        int width, height;
        loader.GoTo("#WIDTH");
        width = loader.NextInt();
        loader.GoTo("#HEIGHT");
        height = loader.NextInt();
        loader.GoTo("#WORLDMODE");
        
        Reset(width, height, WorldMode.valueOf(loader.NextString()));
        
        loader.GoTo("#ORGANISMS_NUMBER");
        int organismsNumber = loader.NextInt();
        
        loader.GoTo("#ORGANISMS");
        for(int i = 0; i < organismsNumber; i++)
        {
            
            Position position;
            if(worldMode_ == WorldMode.SQUARE)
                position = (new Point()).Load(loader);
            else
                position = (new Hex()).Load(loader);
            organismsPO.put(position, manager.LoadOrganism(loader, this, position));
        }
    }
    
    public Organism GetOrganism(Position position)
    {
        return organismsPO.get(position);
    }
    
    public void AddToLogs(String message)
    {
        logSystem_.Log(message);
    }
    
    private Position GetPositionByScreenCoord(java.awt.Point Point)
    {
        if(worldMode_ == WorldMode.SQUARE)
        {
            int x = (Point.x - BOARD_X)/TILE_SIZE;
            int y = (Point.y - BOARD_Y)/TILE_SIZE;
            return new Point(x, y);
        }
             
        int w = (int) (HEX_SIZE * Math.sqrt(3));
        int h = HEX_SIZE * 2;
        int x = Point.x - BOARD_X + w/2;
        int y = Point.y - BOARD_Y + h/2;
        h = 3*h/4;
        
        int row = y/h;
        int column;
        if(row%2 == 1)
            column = (x - w/2)/w;
        else
            column = x / w;
        
        double relY = y - row*h;
        double relX;
        if(row%2 == 1)
            relX = x - column * w - w / 2;
        else
            relX = x - column * w;
        
        double m = Math.tan(Math.toRadians(30));
        double c =  m * w / 2;
        if(relY < (-m * relX) + c)
        {   
            if(row%2 == 0)
                column--;
            row--;
        }
        else if(relY < (m*relX) - c)
        {
            if(row%2 == 1)
                column++;
            row--;
        }
        
        return new Hex(column, row);
    }
}

File: source python/World.py
from Application.Utilities import Colors
from Application.Utilities.Position.Point import Point
from Application.Utilities.Position.Hex import Hex
from Application.Organisms import *
from Application.Requirements import NoRequirements
from Application.Utilities import Savingsystem
import math
import random
import pygame.freetype
import pygame
import tkinter as tk
from tkinter import simpledialog
from enum import Enum


class WorldMode(Enum):

    SQUARE = 1
    HEXAGON = 2


class World:

    BOARD_X = 20
    BOARD_Y = 20
    SQUARE_TILE_SIZE = 20
    HEX_TILE_SIZE = 12

    DEFAULT_COLOR = Colors.SILVER

    def __init__(self, width, height, worldMode, gamePtr):
        self.__gamePtr = gamePtr
        self.Reset(width, height, worldMode)

    def Reset(self, width, height, worldMode):
        self.__width = width
        self.__height = height
        self.__worldMode = worldMode
        self.__organisms = {}
        self.__removalSet = set()

    def IsInBoundary(self, position):
        return position.IsInBoundary(self.__width, self.__height)

    def Act(self):
        organisms = list(self.__organisms.values())
        organisms.sort(key=lambda organism: organism.GetInstanceNumber())
        organisms.sort(key=lambda organism: organism.GetInitiative(), reverse=True)

        for organism in organisms:
            if not organism.IsAlive() or organism in self.__removalSet:
                continue
            organism.Act()

        self.__removalSet.clear()

    def Display(self, window):

        if self.__worldMode == WorldMode.SQUARE:
            font = pygame.freetype.SysFont('calibri', 17)
            baseRect = pygame.Rect(World.BOARD_X, World.BOARD_Y, World.SQUARE_TILE_SIZE, World.SQUARE_TILE_SIZE)
            for i in range(self.__height):
                for j in range(self.__width):
                    rect = baseRect.move(j * World.SQUARE_TILE_SIZE, i * World.SQUARE_TILE_SIZE)

                    position = Point(j, i)
                    if position in self.__organisms:
                        color = self.__organisms.get(position).GetColor()
                    else:
                        color = Colors.SILVER

                    pygame.draw.rect(window, color, rect)
                    pygame.draw.rect(window, Colors.DARK_GRAY, rect, 1)

                    if position in self.__organisms:
                        textRect = font.get_rect(self.__organisms.get(position).GetLetter())
                        font.render_to(window, rect.move((World.SQUARE_TILE_SIZE - textRect.width) / 2, (World.SQUARE_TILE_SIZE - textRect.height) / 2), self.__organisms.get(position).GetLetter())

        elif self.__worldMode == WorldMode.HEXAGON:
            font = pygame.freetype.SysFont('calibri', 17)
            w = int(World.HEX_TILE_SIZE * math.sqrt(3))
            h = World.HEX_TILE_SIZE * 2
            basePolygon = [(-w/2, -h/4), (0, -h/2), (w/2, -h/4), (w/2, h/4), (0, h/2), (-w/2, h/4)]
            polygon = [(0, 0)] * 6

            for i in range(self.__height):
                for j in range(self.__width):
                    xShift = World.BOARD_X + j*w + (w/2 if i % 2 == 1 else 0)
                    yShift = World.BOARD_Y + i*3*h / 4

                    for k in range(6):
                        polygon[k] = (basePolygon[k][0] + xShift, basePolygon[k][1] + yShift)

                    position = Hex(j, i)
                    if position in self.__organisms:
                        color = self.__organisms.get(position).GetColor()
                    else:
                        color = Colors.SILVER

                    pygame.draw.polygon(window, color, polygon)
                    pygame.draw.polygon(window, Colors.DARK_GRAY, polygon, 1)

                    if position in self.__organisms:
                        textRect = font.get_rect(self.__organisms.get(position).GetLetter())
                        font.render_to(window, (xShift - w / 2 + (w - textRect.width) / 2, yShift - h / 4), self.__organisms.get(position).GetLetter())


    def CreateOrganism(self, position, instance):
        if position in self.__organisms:
            return
        self.__organisms[position] = instance.__class__(position, self)

    def SpawnOrganisms(self, animalsCount, plantsCount):
        if animalsCount + plantsCount > self.__width * self.__height:
            return False

        animals = Animal.__subclasses__()
        animals.remove(Human)
        position = self.__GetPosition().GetRandom(self.__width, self.__height)
        self.__organisms[position] = Human(position, self)

        for i in range(animalsCount - 1):
            while True:
                position = self.__GetPosition().GetRandom(self.__width, self.__height)
                if not position in self.__organisms:
                    self.__organisms[position] = animals[random.randint(0, len(animals) - 1)](position, self)
                    break

        plants = Plant.__subclasses__()
        for i in range(plantsCount):
            while True:
                position = self.__GetPosition().GetRandom(self.__width, self.__height)
                if not position in self.__organisms:
                    self.__organisms[position] = plants[random.randint(0, len(plants) - 1)](position, self)
                    break

    def FindNearestPosition(self, position, requirements):
        nearestPosition = None
        lowestDistance = 2 * (self.__width + self.__height)

        if isinstance(requirements, NoRequirements):
            return position.GetRandomNeighbour()

        for i in range(self.__height):
            for j in range(self.__width):
                tmpPosition = self.__GetPosition()(j, i)

                if tmpPosition == position:
                    continue

                currentDistance = position.Distance(tmpPosition)
                if requirements.IsOk(self.__organisms.get(tmpPosition)) and currentDistance < lowestDistance:
                    nearestPosition = tmpPosition
                    lowestDistance = currentDistance

        return nearestPosition

    def GetRandomPlaceInNeighbourhood(self, position, requirements):
        nearestPosition = self.FindNearestPosition(position, requirements)
        if nearestPosition is None or nearestPosition.Distance(position) > 1:
            return None

        while True:
            result = position.GetRandomNeighbour()
            if self.IsInBoundary(result) and requirements.IsOk(self.__organisms.get(result)) and result != position:
                break

        return result

    def MoveInternally(self, organism, position):
        oldPosition = organism.GetPosition()
        if oldPosition == position:
            return

        organism.SetPosition(position)
        self.__organisms.pop(oldPosition, None)
        self.__organisms.pop(position, None)
        self.__organisms[position] = organism

    def MoveAnimal(self, animal, newPosition):
        if not self.IsInBoundary(newPosition):
            print("This should never happen")
            return

        if self.__organisms.get(newPosition) is None or (animal.RunCollision(self.__organisms.get(newPosition)) and animal.IsAlive()):
            self.MoveInternally(animal, newPosition)

    def AddToRemovalSet(self, organism):
        self.__removalSet.add(organism)
        self.__organisms.pop(organism.GetPosition(), None)

    def GetOrganism(self, position):
        return self.__organisms.get(position)

    def GetKey(self):
        return self.__gamePtr.GetKey()

    def __GetPosition(self):
        if self.__worldMode == WorldMode.SQUARE:
            return Point
        elif self.__worldMode == WorldMode.HEXAGON:
            return Hex
        return None

    def HandleClick(self, pos):
        position = self.__GetPositionByScreenCoord(pos)

        if not self.IsInBoundary(position):
            return

        if self.__organisms.get(position) is None:
            window = tk.Tk()
            window.withdraw()

            organisms = Animal.__subclasses__() + Plant.__subclasses__()
            organisms.remove(Human)

            prompt = "Wybierz numer: \n"
            i = 0
            for organismClass in organisms:
                prompt += str(i) + ". " + organismClass.__str__() + "\n"
                i += 1

            answer = simpledialog.askinteger("Tworzenie organizmu", prompt, parent=window, minvalue=0, maxvalue=len(organisms))
            if answer is not None:
                self.__organisms[position] = organisms[answer](position, self)
                self.__gamePtr.Display()
            window.quit()

    def __GetPositionByScreenCoord(self, pos):
        if self.__worldMode == WorldMode.SQUARE:
            return Point(int((pos[0] - World.BOARD_X) / World.SQUARE_TILE_SIZE), int((pos[1] - World.BOARD_Y) / World.SQUARE_TILE_SIZE))
        elif self.__worldMode == WorldMode.HEXAGON:
            w = int(World.HEX_TILE_SIZE * math.sqrt(3))
            h = World.HEX_TILE_SIZE * 2
            x = int(pos[0] - World.BOARD_X + w/2)
            y = int(pos[1] - World.BOARD_Y + h/2)
            h = int(3 * h / 4)

            row = int(y/h)
            if row % 2 == 1:
                column = int((x-int(w/2))/w)
            else:
                column = int(x / w)

            relY = float(y - row*h)
            if row % 2 == 1:
                relX = float(x - column * w - w / 2)
            else:
                relX = float(x - column * w)

            m = float(math.tan(math.radians(30)))
            c = float(m * w / 2)

            if relY < - m*relX + c:
                if row % 2 == 0:
                    column -= 1
                row -= 1
            elif relY < m*relX - c:
                if row % 2 == 1:
                    column += 1
                row -= 1

            return Hex(column, row)

    def Save(self, saver):
        saver.Insert("#WIDTH")
        saver.Insert(self.__width)
        saver.Insert("#HEIGHT")
        saver.Insert(self.__height)
        saver.Insert("#WORLDMODE")
        saver.Insert(self.__worldMode.name)
        saver.Insert("#ORGANISMS_NUMBER")
        saver.Insert(len(self.__organisms))

        saver.Insert("#ORGANISMS")
        for i in range(self.__height):
            for j in range(self.__width):
                position = self.__GetPosition()(j, i)
                if self.__organisms.get(position) is None:
                    continue
                position.Save(saver)
                organism = self.__organisms.get(position)
                saver.Insert(str(organism))
                organism.Save(saver)

    def Load(self, loader):
        loader.GoTo("#WIDTH")
        width = int(loader.Next())
        loader.GoTo("#HEIGHT")
        height = int(loader.Next())
        loader.GoTo("#WORLDMODE")
        worldMode = WorldMode[loader.Next()]
        self.Reset(width, height, worldMode)
        loader.GoTo("#ORGANISMS_NUMBER")
        organismNumber = int(loader.Next())
        loader.GoTo("#ORGANISMS")

        organisms = Animal.__subclasses__() + Plant.__subclasses__()

        for i in range(organismNumber):
            position = self.__GetPosition().Load(loader)
            name = loader.Next()
            for organismClass in organisms:
                if organismClass.__str__() == name:
                    self.__organisms[position] = organismClass(position, self, loader=loader)

        self.__gamePtr.Display()
